### PR TITLE
[dynamo] md5 hash non `compile_ignored` configs

### PR DESCRIPTION
**Pull request by jon-chuang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111300
* #111299
* __->__ #111298
* #111303

fixes: https://github.com/pytorch/pytorch/issues/111235

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @ezyang @Chillee

### CODE REVIEW

File: torch/_dynamo/variables/torch.py [hunk 1]
```diff
@@ -236,15 +236,20 @@ def call_function(
...
+                return value.__module__ + "." + value.__name__
+            return None
+
         if self.value is torch._functorch.vmap.vmap_impl:
             return TorchHigherOrderOperatorVariable.make(
                 self.value,
                 source=self.source,
             ).call_function(tx, args, kwargs)
-        elif self.value in config.constant_functions:
+        elif (full_name := get_full_name(self.value)) in config.constant_functions:
```
Comment (jon-chuang) [inline torch/_dynamo/variables/torch.py:249]: This was to make the config more determinstic. Please advise if this is ok @voznesenskym @ezyang

File: torch/_dynamo/config.py [hunk 2]
```diff
@@ -157,13 +156,13 @@
 # the `allowed_functions.is_allowed` function will not consider it
 # when creating a list of PyTorch functions that will appear in
 # FX IR.
-allowed_functions_module_string_ignorelist = {
+allowed_functions_module_string_ignorelist = [
```
Comment (jon-chuang) [inline torch/_dynamo/config.py:159]: This was to make the config more determinstic. Please advise if this is ok @voznesenskym @ezyang

File: torch/_dynamo/config.py [hunk 3]
```diff
@@ -50,13 +49,13 @@
...
 # Assume these functions return constants
 constant_functions = {
-    torch.jit.is_scripting: False,
-    torch.jit.is_tracing: False,
-    torch._C._get_tracing_state: None,
-    torch.fx._symbolic_trace.is_fx_tracing: False,
-    torch.onnx.is_in_onnx_export: False,
-    external_utils.is_compiling: True,
-    torch._utils.is_compiling: True,
+    "torch.jit.is_scripting": False,
```
Comment (jon-chuang) [inline torch/_dynamo/config.py:52]: This was to make the config hash more determinstic. Please advise if this is ok @voznesenskym @ezyang

File: torch/_dynamo/variables/torch.py [hunk 4]
```diff
@@ -236,15 +236,21 @@ def call_function(
         unspec_python_args = check_unspec_python_args(args, kwargs)
         options = VariableTracker.propagate(self, args, kwargs.values())
 
+        def get_full_name(value):
+            if hasattr(value, "__module__") and hasattr(value, "__name__"):
+                if value.__module__ is not None and value.__name__ is not None:
+                    return value.__module__ + "." + value.__name__
```
Comment (jon-chuang) [inline torch/_dynamo/variables/torch.py:242]: nit: you could do `f"{value.__module__ }.{value.__name__}"` instead?

File: torch/_dynamo/config_utils.py [hunk 5]
```diff
@@ -177,6 +182,14 @@ def codegen_config(self):
             lines.append(f"{mod}.{k} = {v!r}")
         return "\n".join(lines)
 
+    def get_hash(self):
+        """Hashes the configs that are not compile_ignored"""
+        if self._is_dirty or self._hash_digest is None:
+            string_to_hash = repr(sorted(self._config.items()))
+            self._hash_digest = hashlib.md5(string_to_hash.encode("utf-8")).digest()
```
Comment (jon-chuang) [inline torch/_dynamo/config_utils.py:189]: hmm, I think I finally decided on `.hexdigest()`, please change it to that.

Comment (jon-chuang) [reply torch/_dynamo/config_utils.py:189]: Actually, maybe let's just do `.digest()` and `hex()` it

File: torch/_dynamo/variables/torch.py (hunk 4 above)
Comment (jon-chuang) [reply torch/_dynamo/variables/torch.py:242]: I think this change is too risky after all. Let's put it as a TODO in a standalone PR.

File: torch/_dynamo/config.py (hunk 3 above)
Comment (jon-chuang) [reply torch/_dynamo/config.py:52]: I think this change is too risky after all. Let's put it as a TODO in a standalone PR.

File: torch/_dynamo/config.py (hunk 2 above)
Comment (jon-chuang) [reply torch/_dynamo/config.py]: Relocated to new PR

**Comment by jon-chuang** [comment]
Ready for review

**Review by ezyang — APPROVED** [review]

**Comment by jon-chuang** [comment]
Can't seem to merge. What do I need to do @ezyang @voznesenskym

**Comment by jon-chuang** [comment]
Oh wait, I think pytorch bot merged but just didn't close the PR. lol

**Pull request closed without merging**